### PR TITLE
Fix S1244 FN: Half, NFloat, and IFloatingPointIeee754 are not supported

### LINKED ISSUE #6423: Fix S1244 FN: Half, NFloat, and IFloatingPointIeee754 are not supported
### Description

All types deriving from [IFloatingPointIeee754](https://learn.microsoft.com/en-us/dotnet/api/system.numerics.ifloatingpointieee754-1?view=net-7.0) suffer from equivalency check problems. S1244 needs to be adopted to take these into account.

Test cases added in #6422

### Open question

We did not support the decimal type in S1244, and we probably shouldn't in the future. Therefore we should not take IFloatingPoint into account. Inheritance hierarchy:
* IFloatingPoint (e.g., decimal, half, single, double, etc.) Base type for all decimal digit types
* IFloatingPointIeee754 (e.g., half, single, double, etc.) Decimal is missing from the list as it uses 10-based numbers for representing instead of 2 based.
* IBinaryFloatingPointIeee754 (all .Net IFloatingPointIeee754 are also IBinaryFloatingPointIeee754)

It is a bit confusing that [IEEE 754 - Wikipedia](https://en.wikipedia.org/wiki/IEEE_754) defines base 2 and base 10 encoded formats, so it is unclear why the .Net type decimal is not an IFloatingPointIeee754. Some research is needed to clarify which interface is appropriate to test for (IFloatingPoint or IFloatingPointIeee754).

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Fixes #6423

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs [hunk 1]
```diff
@@ -5,14 +5,24 @@
 public class EqualityOnFloatingPoint
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs:5]: From looking at the tests and implementation, I guess we don't support

```
if (f <= 3.146 && 3.146 <= f) // FN?
```
?

Could you maybe add this test to have it documented in the initial non C# 11 test cases?

I wouldn't fix it in this PR to avoid growing it too much (and it's out of scope for this issue & PR)

eventually open a separate ticket and fix it in a separate PR?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs [hunk 2]
```diff
@@ -5,14 +5,24 @@
...
+        => first == second;    // Noncompliant
 
     bool IsEpsilon<T>(T value) where T : IFloatingPointIeee754<T>
-        => value == T.Epsilon; // FN
+        => value == T.Epsilon; // Noncompliant
+
+    bool IsPi<T>(T value) where T : IFloatingPointIeee754<T>
+        => value <= T.Pi && ((value >= T.Pi)); // Noncompliant
+
+    bool IsNotE<T>(T value) where T : IFloatingPointIeee754<T>
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs:20]: what about 

- [NaN](https://learn.microsoft.com/en-us/dotnet/api/system.numerics.ifloatingpointieee754-1.nan?view=net-7.0#system-numerics-ifloatingpointieee754-1-nan)
- [NegativeInfinity](https://learn.microsoft.com/en-us/dotnet/api/system.numerics.ifloatingpointieee754-1.negativeinfinity?view=net-7.0#system-numerics-ifloatingpointieee754-1-negativeinfinity)
- [PositiveInfinity](https://learn.microsoft.com/en-us/dotnet/api/system.numerics.ifloatingpointieee754-1.positiveinfinity?view=net-7.0#system-numerics-ifloatingpointieee754-1-positiveinfinity)

should equality be treated the same as Noncompliant? I imagine not?

**Review by andrei-epure-sonarsource — CHANGES_REQUESTED** [review]
first commenst

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs (hunk 2 above)
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs:20]: From the docs:

> Use the [IsNaN](https://learn.microsoft.com/en-us/dotnet/api/system.double.isnan?view=net-7.0) method to determine whether a value is not a number. The [Equality](https://learn.microsoft.com/en-us/dotnet/api/system.double.op_equality?view=net-7.0) operator considers two [NaN](https://learn.microsoft.com/en-us/dotnet/api/system.double.nan?view=net-7.0) values to be unequal to one another.

[Source](https://learn.microsoft.com/en-us/dotnet/api/system.double.nan?view=net-7.0#remarks)

For the others:

> Use [IsPositiveInfinity](https://learn.microsoft.com/en-us/dotnet/api/system.double.ispositiveinfinity?view=net-7.0) to determine whether a value evaluates to positive infinity.

> Use [IsNegativeInfinity](https://learn.microsoft.com/en-us/dotnet/api/system.double.isnegativeinfinity?view=net-7.0) to determine whether a value evaluates to negative infinity.

There is nothing mentioned for NegativeZero.

The "Is..." methods are defined in [INumberBase](https://learn.microsoft.com/en-us/dotnet/api/system.numerics.inumberbase-1.isnan?view=net-7.0#system-numerics-inumberbase-1-isnan(-0)) and IFloatingPointIeee754 derives from INumberBase.

The only exception seems to be NegativeZero which doesn't have an `IsNegativeZero` counterpart.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs [hunk 3]
```diff
@@ -5,14 +5,24 @@
...
+        => value > T.E || ((value < T.E)); // Noncompliant
 
     bool AreEqual<T>(T first, T second) where T : IFloatingPointIeee754<T>
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs:23]: Another good reason is `Epsilon`. This value is tightly coupled with the equality problem (see e.g. [Single.Equals Method (System)](https://learn.microsoft.com/en-us/dotnet/api/system.single.equals?view=net-7.0). It is defined for [IFloatingPointIeee754<TSelf>.Epsilon Property (System.Numerics)](https://learn.microsoft.com/en-us/dotnet/api/system.numerics.ifloatingpointieee754-1.epsilon?view=net-7.0) but not for IFloatingPoint

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs [hunk 4]
```diff
@@ -64,80 +64,66 @@ private static void CheckLogicalExpression(SyntaxNodeAnalysisContext context)
...
-        private static bool HasAppropriateOperatorsForInequality(BinaryExpressionSyntax right, BinaryExpressionSyntax left)
-        {
-            return new[] { right.OperatorToken.Kind(), left.OperatorToken.Kind() }
-                .Intersect(new[] { SyntaxKind.LessThanToken, SyntaxKind.GreaterThanToken })
-                .Count() == 2;
-        }
+        private static bool IsExpressionFloatingType(SemanticModel semanticModel, ExpressionSyntax expression) =>
+            IsFloatingPointNumberType(semanticModel.GetTypeInfo(expression).Type);
+
+        private static bool HasAppropriateOperatorsForEquality(BinaryExpressionSyntax right, BinaryExpressionSyntax left) =>
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs:121]: I didn't realize the order before but it is confusing indeed and I changed it.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs (hunk 1 above)
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs:5]: Good catch. I added the test. If you don't mind, I will not open a ticket for this. I think the whole indirect equality check is of questionable value. It would be a better idea to have a separate rule that finds possible simplifications.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs:23]: I added a code comment to my best current knowledge explaining the reasons for choosing that interface and reasons why decimal is excluded.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs [hunk 5]
```diff
@@ -64,80 +64,71 @@ private static void CheckLogicalExpression(SyntaxNodeAnalysisContext context)
...
-        {
-            return new[] { right.OperatorToken.Kind(), left.OperatorToken.Kind() }
-                .Intersect(new[] { SyntaxKind.LessThanToken, SyntaxKind.GreaterThanToken })
-                .Count() == 2;
-        }
+        // All floating point types that suffer from equivalence problems. These are all .net floating point types except decimal.
+        // Reason: The Equals documentation for decimal does not have a "Precision in Comparisons" section as the other floating point types.
+        // Power 2 based types like double implement IFloatingPointIeee754 but power 10 based decimal not (implements IFloatingPoint).
+        // Ieee754 also allows power 10 based representations but uses another representation than the .Net decimal type.
+        // IFloatingPointIeee754 defines Epsilon which indicates problems with equivalence checking.
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/EqualityOnFloatingPoint.cs:98]: I suggest some simplifications/improvements.

I find the `Ieee754` mention confusing, and I'd rather avoid mentioning it.

```suggestion
        // Returns true for the floating point types that suffer from equivalence problems. All .NET floating point types have this problem except `decimal.`
        // -  the documentation for the `decimal.Equals()` method does not have a "Precision in Comparisons" section as the other .NET floating point types.
        // - Power-2-based types like `double` implement `IFloatingPointIeee754`, but power-10-based `decimal` implements `IFloatingPoint`.
        // - `IFloatingPointIeee754` defines `Epsilon` which indicates problems with equivalence checking.
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs (hunk 2 above)
Comment (andrei-epure-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs:20]: Fair point. In this case, the rule message will be confusing
```
Do not check floating point equality with exact values, use a range instead.
```
because for these values there are clear framework provided methods for doing the checks:
- IsNaN
- IsPositiveInfinity
- IsNegativeInfinity

And the rule should rather say that in the message. I'd say that there should be a ticket for this improvement.

**Review by andrei-epure-sonarsource — APPROVED** [review]
LGTM

I suggest improving the overall comment and opening a ticket before merging.

**Review by andrei-epure-sonarsource — CHANGES_REQUESTED** [review]
I've noticed now that the code coverage dropped vs [the master](https://sonarcloud.io/code?id=sonaranalyzer-dotnet&selected=sonaranalyzer-dotnet%3Aanalyzers%2Fsrc%2FSonarAnalyzer.CSharp%2FRules%2FEqualityOnFloatingPoint.cs) (probably because replacing `Union`)

It should be possible without a big effort to add the test cases with the right combinations to get back to 100%?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs (hunk 2 above)
Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/EqualityOnFloatingPoint.CSharp11.cs:20]: #6521
